ISPN-10539 Server loggers management

tristantarrant opened the pull request:
https://issues.redhat.com/browse/ISPN-10539
https://issues.redhat.com/browse/ISPN-11238
https://issues.redhat.com/browse/ISPN-11477

pruivo commented on cli/cli-client/src/main/java/org/infinispan/cli/connection/rest/RestConnection.java:
```diff
@@ -361,7 +361,7 @@ public String execute(List<CommandInputLine> commands) throws IOException {
                break;
             }
-            case Ls.CMD: {
+            case org.infinispan.cli.commands.Ls.CMD: {
```
why?

pruivo commented on client/rest-client/src/main/java/org/infinispan/client/rest/RestServerClient.java:
```diff
@@ -36,4 +36,6 @@
 
    CompletionStage<RestResponse> listIgnoredCaches(String cacheManagerName);
+
+   RestLoggingClient logging();
```
why in `RestServerClient`? Shouldn't the logging affect all nodes in the cluster? 
maybe under `RestClient` or `RestClusterClient`?

pruivo commented on client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestLoggingClientOkHttp.java:
```diff
@@ -0,0 +1,66 @@
+   @Override
+   public CompletionStage<RestResponse> removeLogger(String name) {
+      Request.Builder builder = new Request.Builder();
+      builder.url(baseLoggingURL + "/" + sanitize(name)).delete();
```
isn't `"/loggers/"` missing?

tristantarrant replied:
Fixed

pruivo commented on client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestLoggingClientOkHttp.java:
```diff
@@ -0,0 +1,66 @@
+         amp = true;
+      }
+      if (appenders != null) {
+         for(String appender : appenders) {
```
nitpick: `sb.append("appender=").append(String.join("%appender=", appenders))`

pruivo commented on documentation/src/main/asciidoc/topics/proc_cli_logging.adoc:
```diff
@@ -0,0 +1,93 @@
+
+The above command will return:
+
+[source,javascript]
```
```
[source,json,options="nowrap",subs=attributes+]
----
include::json_examples/rest_logging_appenders_response.json[]
----
```

tristantarrant replied:
Done

pruivo commented on documentation/src/main/asciidoc/topics/proc_cli_logging.adoc:
```diff
@@ -0,0 +1,93 @@
+
+The above command will return:
+
+[source,javascript]
```
```
[source,json,options="nowrap",subs=attributes+]
----
include::json_examples/rest_logging_loggers_response.json[]
----
```

tristantarrant replied:
Done

pruivo commented on server/rest/src/main/java/org/infinispan/rest/resources/LoggingResource.java:
```diff
@@ -0,0 +1,180 @@
+      public void serialize(Appender appender, JsonGenerator json, SerializerProvider serializerProvider) throws IOException {
+         json.writeStartObject();
+         json.writeStringField("name", appender.getName());
+         Layout<? extends Serializable> layout = appender.getLayout();
```
layout isn't used.

tristantarrant replied:
Done

pruivo commented on server/runtime/src/main/java/org/infinispan/server/tasks/admin/LoggingSetTask.java:
```diff
@@ -0,0 +1,93 @@
+   private final Set<String> PARAMETERS;
+
+   public LoggingSetTask() {
+      PARAMETERS = Stream.of("loggerName", "level", "appenders", "create").collect(Collectors.toSet());
```
static?

tristantarrant replied:
Done

pruivo commented on server/tests/src/test/java/org/infinispan/server/functional/RestLoggingResource.java:
```diff
@@ -0,0 +1,85 @@
+   }
+
+   @Test
+   public void testManipulateLogger() throws Exception {
```
can you duplicate the test using `RestLoggingClient`? It would avoid issues in `RestLoggingClient` :)

tristantarrant replied:
Done

tristantarrant commented on cli/cli-client/src/main/resources/help/logging.adoc:
```diff
@@ -0,0 +1,38 @@
+
+
+LOGGING SET OPTIONS
+--------------------
```
Done

tristantarrant commented on cli/cli-client/src/main/resources/help/logging.adoc:
```diff
@@ -0,0 +1,38 @@
+
+*logging set ['OPTIONS'] `LOGGER_NAME`*
+
+*logging remove `LOGGER_NAME`*
```
Done

tristantarrant commented on cli/cli-client/src/main/resources/help/logging.adoc:
```diff
@@ -0,0 +1,38 @@
+
+*logging list-appenders*
+
+*logging set ['OPTIONS'] `LOGGER_NAME`*
```
Done

tristantarrant commented on cli/cli-client/src/main/resources/help/logging.adoc:
```diff
@@ -0,0 +1,38 @@
+Specifies the logging level for the specific logger.
+
+*-a, --appender*='APPENDER'::
+Specifies one or more appenders to set on the specific logger.
```
Done

tristantarrant commented on cli/cli-client/src/main/java/org/infinispan/cli/commands/CommandInputLine.java:
```diff
@@ -80,6 +80,9 @@ public long longOption(String option) {
    public boolean boolOption(String option) {
       return (Boolean) options.get(option);
    }
+   public <T> T optionAs(String option) {
```
Done

gustavonalle commented on server/rest/src/main/java/org/infinispan/rest/resources/LoggingResource.java:
```diff
@@ -0,0 +1,177 @@
+   @Override
+   public Invocations getInvocations() {
+      return new Invocations.Builder()
+            .invocation().methods(GET).path("/v2/logging/loggers").handleWith(this::listLoggers)
```
This resource will only affect the local server, correct? I am wondering if it should be mapped under ```/v2/server/logging```

tristantarrant replied:
Tricky, because the set and remove ops are clustered

tristantarrant replied:
So `GET /v2/server/logging/loggers` and `GET /v2/server/logging/appenders` but `PUT /v2/cluster/logging/loggers/org.infinispan.ABC` and `DELETE /v2/cluster/logging/loggers/org.infinispan.ABC`

gustavonalle replied:
WDYM? PUT and DELETE are sent to all nodes, but not GET?

tristantarrant replied:
PUT and DELETE are handled by clustered tasks.
GET is only handled by the local node. In a symmetrical cluster that should be fine. Making it clustered would mean returning a map of list of loggers (one list per server) which seems overkill

tristantarrant commented:
@pruivo fixed the issue you were having. Also adds completers for levels and appenders

pruivo commented:
it is working for me now. 
@tristantarrant have you had the time to check https://gist.github.com/pruivo/81d04e44027fd27ea840dd13486c5190 ?

tristantarrant commented:
@pruivo answers to your Gist:
Q: com.arjuna doesn't have any appenders. bug?
A: That's what Log4j reports. It  means it will use the appenders of the root logger
Q: is it possible to use some defaults? or get the level/appenders from the root
A: I've added a tab completer which will complete with existing loggers.
Q: no appenders? 
A: See answer above.
Q:validate level input? It is a no-op but an error should be shown IMO.
Q: It seems to use DEBUG as default. Can it pick the parent's level?
A: that's Log4j's behaviour. I've changed it to use the root level if unspecified and to print an error if it's unknown.
Q: isn't this suppose to work?
A: I've created the org.infinispan logger and removed it successfully. 
Q: Suggestion: print the logger details after set. Example:
A: none of our RESTful PUT methods return the value and I decided for consistency

pruivo commented on server/testdriver/junit5/pom.xml:
```diff
@@ -0,0 +1,54 @@
+      <relativePath>../../pom.xml</relativePath>
+   </parent>
+
+   <artifactId>infinispan-server-testdriver-junit4</artifactId>
```
remove? it doesn't belong here. (I think)

tristantarrant replied:
Fixed.

pruivo commented:
integrated! thanks @tristantarrant !

gustavonalle commented on documentation/src/main/asciidoc/topics/ref_rest_logging.adoc:
```diff
@@ -0,0 +1,57 @@
+
+[source,options="nowrap",subs=attributes+]
+----
+GET /rest/v2/logging/appenders
```
It should be documented that it will only get the logger from the local server, or move the API to under /server

gustavonalle commented on documentation/src/main/asciidoc/topics/ref_rest_logging.adoc:
```diff
@@ -0,0 +1,57 @@
+
+[source,options="nowrap",subs=attributes+]
+----
+PUT /rest/v2/logging/loggers/{loggerName}&level={level}&appender={appender}&appender={appender}...
```
It should be document that it will affect all servers in the cluster

gustavonalle commented on documentation/src/main/asciidoc/topics/ref_rest_logging.adoc:
```diff
@@ -0,0 +1,57 @@
+
+[source,options="nowrap",subs=attributes+]
+----
+DELETE /rest/v2/logging/loggers/{loggerName}
```
Same as above, it should be document that it will affect all servers in the cluster
